test/cpp/tensorexpr/test_memplanning.cpp

```diff
@@ -85,6 +87,253 @@ TEST(BufLiveRange, MulRangeLine) {
+      EP, {i, j}, Cast::make(kQUInt8, DP.load(i, j) + DP.load(i, j)));
+  auto loop_e = For::make(i, 0, M, For::make(j, 0, N, store_e));
+
+  auto store_f = Store::make(FP, {i, j}, Load::make(EP, {i, j}));
+  auto loop_f = For::make(i, 0, M, For::make(j, 0, N, store_f));
+
+  auto stmt = Block::make({loop_c, loop_d, loop_e, loop_f});
+  // Constructed stmt:
+  // Intermediate buffers and their liveness ranges: C [0, 1], D [1, 2],
+  // E [2, 3]. The dimensions of 'C' and 'E' are the same but their types are
+  // different: 'E' type quint8 < 'C' type float. We'll reuse 'C' for 'E'
+  // with typecasting.
+  // {
+  //   for (int i = 0; i < 32; i++) {
+  //     for (int j = 0; j < 32; j++) {
+  //       C[i, j] = float(0);
+  //       for (int k = 0; k < 32; k++) {
+  //         C[i, j] = ReduceOp((C[i, j]) + (A[i, k]) * (B[k, j]),
+  //         reduce_args={k});
+  //       }
+  //     }
+  //   }
+  //   for (int i = 0; i < 32; i++) {
+  //     for (int j = 0; j < 32; j++) {
+  //       D[i, j] = (C[i, j])<float(0) ? float(0) : (C[i, j]);
+  //     }
+  //   }
+  //   for (int i = 0; i < 32; i++) {
+  //     for (int j = 0; j < 32; j++) {
+  //       E[i, j] = quint8((D[i, j]) + (D[i, j]));
+  //     }
+  //   }
+  //   for (int i = 0; i < 32; i++) {
+  //     for (int j = 0; j < 32; j++) {
+  //       F[i, j] = E[i, j];
+  //     }
+  //   }
+  // }
+
+  LoopNest l(stmt, {FP.node()});
+  l.prepareForCodegen();
+  SimpleIREvaluator cg(Stmt::clone(l.root_stmt()), {AP, BP, FP});
+
+  checkIR(cg.stmt(), R"IR(
+# CHECK: Allocate(C); // dtype=float, dims=[32, 32]
+# CHECK: Allocate(D); // dtype=float, dims=[32, 32]
+# CHECK: Alias(E,C);
+# CHECK: Free(D);
+# CHECK: Free(C))IR");
+
+  PaddedBuffer<float> a_v(M, K, "a");
+  PaddedBuffer<float> b_v(K, N, "b");
+  PaddedBuffer<uint8_t> o1(M, N, "e_before");
+  PaddedBuffer<uint8_t> o2(M, N, "e_after");
+
+  for (const auto m : c10::irange(M)) {
+    for (const auto k : c10::irange(K)) {
+      a_v(m, k) = at::randn({1}).item().to<float>();
+    }
+  }
+
+  for (const auto k : c10::irange(K)) {
+    for (const auto n : c10::irange(N)) {
+      b_v(k, n) = at::randn({1}).item().to<float>();
+    }
+  }
+
+  cg.call({a_v, b_v, o1});
+
+#ifdef TORCH_ENABLE_LLVM
+  LLVMCodeGen cg_llvm(Stmt::clone(l.root_stmt()), {AP, BP, FP});
+
+  checkIR(cg_llvm.stmt(), R"IR(
+# CHECK: Allocate(C); // dtype=float, dims=[32, 32]
+# CHECK: Allocate(D); // dtype=float, dims=[32, 32]
+# CHECK: Alias(E,C);
+# CHECK: Free(D);
+# CHECK: Free(C))IR");
+
+  cg_llvm.call({a_v, b_v, o2});
+
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+  ExpectAllNear(o1, o2, 1e-5);
+#endif
+}
+
+TEST(MemPlanning, NoMemReuseForLargerType) {
+  int M = 32;
+  int N = 32;
+  int K = 32;
+
+  BufHandle AP("A", {M, K}, kShort);
+  BufHandle BP("B", {K, N}, kShort);
+  BufHandle CP("C", {M, N}, kShort);
+  BufHandle DP("D", {M, N}, kShort);
+  BufHandle EP("E", {M, N}, kFloat);
+  BufHandle FP("F", {M, N}, kFloat);
```

Comment:
This test is almost identical to the previous one. Can we refactor them to avoid code duplication?

Reply:
The stmt construction process is reusable. I tried making it a common process for the two tests but the code is ugly because we need both the constructed stmt and the input/output buffers for checking. The constructed stmt is checked for `Alloc/PlacementAlloc/Free` stmts and the input/output buffers are needed for codegen construction which verifies the results. The pack/unpack of stmt + output buffer is ugly, and the reuse does not save that many lines. So will keep it as two separate tests.